tensorflow-core/tensorflow-core-api/src/main/java/org/tensorflow/op/core/NN.java

```diff
@@ -0,0 +1,379 @@
+    if (logitsShape.numDimensions() == 0) {
+      throw new IllegalArgumentException(
+          String.format("Logits cannot be scalars - received shape %s.", logitsShape));
+    }
+    if (!logitsShape.hasUnknownDimension()
+        && !labelsStaticShape.hasUnknownDimension()
+        && labelsStaticShape.numDimensions() != logitsShape.numDimensions() - 1) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Rank mismatch: Rank of labels (received %s) should equal rank of logits minus 1 (received %s).",
+              labelsStaticShape.toString(), logitsShape.toString()));
+    }
+
+    if (staticShapesFullyDefined && !labelsStaticShape.equals(logitsShortened)) {
+      throw new IllegalArgumentException(
+          String.format(
+              "Shape mismatch: The shape of labels (received %s) "
+                  + "should equal the shape of logits except for the last "
+                  + "dimension (received %s).",
+              labelsStaticShape.toString(), logitsShape.toString()));
+    }
+    // Check if no reshapes are required.
+    if (logitsShape.numDimensions() == 2) {
+      SparseSoftmaxCrossEntropyWithLogits smax =
+          SparseSoftmaxCrossEntropyWithLogits.create(scope, precise_logits, labels);
+      Operand loss = smax.loss();
+      if (logits.asOutput().dataType() == TFloat16.DTYPE) {
+        loss = Cast.create(scope, loss, TFloat16.DTYPE);
+      }
+      return loss;
+    }
+
+    List<Op> shapeChecks = new ArrayList<>();
+
+    if (!staticShapesFullyDefined) {
+      shapeChecks.add(
+          AssertThat.create(
+              scope,
+              Equal.create(
+                  scope,
+                  org.tensorflow.op.core.Shape.create(scope, labels),
+                  Shapes.take(
+                      scope,
+                      org.tensorflow.op.core.Shape.create(scope, logits),
+                      Constant.scalarOf(scope, -1))),
+              Collections.singletonList(
+                  Constant.scalarOf(
+                      scope,
+                      "Shape mismatch: The shape of labels  "
+                          + "should equal the shape of logits except for the last "
+                          + "dimension "))));
+    }
+
+    // Reshape logits to 2 dim, labels to 1 dim.
+    long numClassses = logitsShape.size(logitsShape.numDimensions() - 1);
+
+    precise_logits =
+        Reshape.create(
+            scope, precise_logits, Constant.vectorOf(scope, new long[] {-1, numClassses}));
+    labels = Reshape.create(scope, labels, Constant.scalarOf(scope, -1));
+    scope.withControlDependencies(shapeChecks);
+    SparseSoftmaxCrossEntropyWithLogits smax =
+        SparseSoftmaxCrossEntropyWithLogits.create(scope, precise_logits, labels);
+    Operand cost = smax.loss();
+    cost = Reshape.create(scope, cost, labelsShape);
+    if (logits.asOutput().dataType() == TFloat16.DTYPE) {
+      cost = Cast.create(scope, cost, TFloat16.DTYPE);
```

Comment:
I think you are missing `TBfloat16` here?

Reply:
Python source only refers to dtypes.float16.  It does not show anything with dtypes.bfloat16.
```
cost, _ = gen_nn_ops.sparse_softmax_cross_entropy_with_logits(precise_logits, labels, name=name)
if logits.dtype == dtypes.float16:
        return math_ops.cast(cost, dtypes.float16)
else:
        return cost
```